src/lib/salad/hll.c

```diff
@@ -0,0 +1,223 @@
+			free(hll);
+			return NULL;
+		}
+	}
+	return hll;
+}
+
+void
+hll_destroy(struct hll *hll)
+{
+	free(hll->registers);
+	free(hll);
+}
+
+/*
+ * The highest precision bits of the hash are interpreted as a register index.
+ */
+static inline size_t
+hll_hash_register_idx(uint64_t hash, uint8_t precision)
+{
+	return hash >> (64 - precision);
+}
+
+/*
+ * Get a number whose first n bits are equal to ones.
+ */
+static inline uint64_t
+hll_ones(size_t n)
+{
+	return ((1ULL << n) - 1);
+}
+
+/*
+ * Return the number of leading zeros of the first
+ * (64 - precision) hash bits plus one.
+ */
+static inline uint8_t
+hll_hash_rank(uint64_t hash, uint8_t precision)
+{
+	hash |= hll_ones(precision) << (64 - precision);
+	uint8_t zero_count = 0;
+	uint64_t bit = 0x1;
+	while ((hash & bit) == 0) {
+		++zero_count;
+		bit <<= 1;
+	}
+	assert(zero_count < (1u << HLL_REGISTER_BITS));
+	return zero_count + 1;
+}
+
+/*
+ * HyperLogLog register is represented by 6 bits so it can go out
+ * the range of one byte but it always fits in a 2 bytes bucket.
+ */
+struct reg_bucket {
+	/* Pointer to bucket where register is stored. */
+	uint16_t *addr;
+	/* Offset of the register in the bucket. */
+	size_t offset;
+};
+
+/*
+ * Init register bucket structure that is used for
+ * convenient work with registers.
+ */
+static inline void
+init_reg_bucket(struct reg_bucket *bucket, uint8_t *regs, size_t idx)
+{
+	size_t first_byte_pos = idx * HLL_REGISTER_BITS / 8;
+	bucket->addr = (uint16_t *)(regs + first_byte_pos);
+	bucket->offset = idx * HLL_REGISTER_BITS - first_byte_pos * 8;
+}
+
+/* Get the value of the register with the idx index. */
+static inline uint8_t
+hll_register_rank(const struct hll *hll, size_t idx)
+{
+	struct reg_bucket bucket;
+	init_reg_bucket(&bucket, hll->registers, idx);
+	uint16_t reg_mask = hll_ones(HLL_REGISTER_BITS);
+	uint16_t bucket_val = *bucket.addr;
+	uint8_t rank = (bucket_val >> bucket.offset) & reg_mask;
+	return rank;
+}
+
+/* Set rank as the new value of the register with the idx index. */
+static inline void
+hll_set_register_rank(struct hll *hll, size_t idx, uint8_t rank)
+{
+	struct reg_bucket bucket;
+	init_reg_bucket(&bucket, hll->registers, idx);
+	uint16_t boundary_mask =
+		~(hll_ones(HLL_REGISTER_BITS) << bucket.offset);
+	uint16_t bucket_val = *bucket.addr;
+	*bucket.addr = (rank << bucket.offset) | (bucket_val & boundary_mask);
+}
+
+void
+hll_add(struct hll *hll, uint64_t hash)
+{
+	uint8_t precision = hll->precision;
+	size_t idx = hll_hash_register_idx(hash, precision);
+	assert(idx < hll_n_registers(precision));
+	uint8_t hash_rank = hll_hash_rank(hash, precision);
+	uint8_t reg_rank = hll_register_rank(hll, idx);
+	if (reg_rank < hash_rank) {
+		hll_set_register_rank(hll, idx, hash_rank);
+		hll_invalidate_cache(hll);
+	}
+}
+
+/* Alpha constant that HyperLogLog uses for estimating. */
+static inline double
+hll_alpha(uint8_t precision)
+{
+	return 0.7213 / (1.0 + 1.079 / hll_n_registers(precision));
```

Comment:
Please elaborate what this constant is used for.

Reply:
Done.